### DISCUSSION TITLE
How to apply clang-tidy auto-fix (e.g. modernize-* checks)

### DISCUSSION THREAD

**Discussion by FrancescoBorzi** [azerothcore/azerothcore-wotlk#9970] [Guides & Tips]
[clang-tidy](https://clang.llvm.org/extra/clang-tidy/checks/list.html) is

> a clang-based C++ “linter” tool. Its purpose is to provide an extensible framework for diagnosing and fixing typical programming errors, like style violations, interface misuse, or bugs that can be deduced via static analysis. 

Among all its [available checks](https://clang.llvm.org/extra/clang-tidy/checks/list.html), there is a collection of `modernize-*` checks that offer automated fixes. It would be nice to apply them to our codebase in order to make it more "modern" and resolve potential issues and warnings.

[This way](https://www.kdab.com/clang-tidy-part-1-modernize-source-code-using-c11c14/) it is possible to run `clang-tidy` and automatically fix the whole codebase.

After this is done, it would be nice to add an automated check to our CI to prevent new issues to get to master, this is part of https://github.com/azerothcore/azerothcore-wotlk/issues/2745

Current progress:

- [x] **modernize-avoid-bind** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/66e65c474a28900fe722ab790b5a2babf293852a)
- [x] **modernize-concat-nested-namespaces** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/66e65c474a28900fe722ab790b5a2babf293852a)
- [x] **modernize-deprecated-headers** (https://github.com/azerothcore/azerothcore-wotlk/pull/3821)
- [x] **modernize-deprecated-ios-base-aliases** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/66e65c474a28900fe722ab790b5a2babf293852a)
- [x] **modernize-loop-convert** (https://github.com/azerothcore/azerothcore-wotlk/pull/3822)
- [x] **modernize-make-shared** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/66e65c474a28900fe722ab790b5a2babf293852a)
- [x] **modernize-make-unique** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/66e65c474a28900fe722ab790b5a2babf293852a)
- [x] **modernize-pass-by-value** (https://github.com/azerothcore/azerothcore-wotlk/pull/3823)
- [x] **modernize-raw-string-literal** (https://github.com/azerothcore/azerothcore-wotlk/pull/3824)
- [x] **modernize-redundant-void-arg** (https://github.com/azerothcore/azerothcore-wotlk/pull/3825)
- [x] **modernize-replace-auto-ptr** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/0baecd3e8f65db05a252cba8056360fd56b7cf8c)
- [x] **modernize-replace-disallow-copy-and-assign-macro** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/0baecd3e8f65db05a252cba8056360fd56b7cf8c)
- [x] **modernize-replace-random-shuffle** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/0baecd3e8f65db05a252cba8056360fd56b7cf8c)
- [x] **modernize-return-braced-init-list** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/0baecd3e8f65db05a252cba8056360fd56b7cf8c)
- [x] **modernize-shrink-to-fit** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/0baecd3e8f65db05a252cba8056360fd56b7cf8c)
- [x] **modernize-unary-static-assert** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/0baecd3e8f65db05a252cba8056360fd56b7cf8c)
- [ ] modernize-use-auto (this is probably going to be very big, so better do it after the others are merged)
- [x] **modernize-use-bool-literals** (https://github.com/azerothcore/azerothcore-wotlk/pull/3826)
- [x] **modernize-use-default-member-init** (https://github.com/azerothcore/azerothcore-wotlk/pull/3827)
- [x] **modernize-use-emplace** (https://github.com/azerothcore/azerothcore-wotlk/pull/3828)
- [x] **modernize-use-equals-default** (https://github.com/azerothcore/azerothcore-wotlk/pull/3834)
- [x] **modernize-use-equals-delete** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/221f33b5b7af4ebe1b331da167391c10b15622f9)
- [x] **modernize-use-nodiscard** (https://github.com/azerothcore/azerothcore-wotlk/pull/3835)
- [x] **modernize-use-noexcept** (https://github.com/azerothcore/azerothcore-wotlk/pull/3837)
- [x] **modernize-use-nullptr** (https://github.com/azerothcore/azerothcore-wotlk/pull/3819)
- [x] **modernize-use-override** (https://github.com/azerothcore/azerothcore-wotlk/pull/3817)
- [x] **modernize-use-trailing-return-type** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/221f33b5b7af4ebe1b331da167391c10b15622f9)
- [x] **modernize-use-transparent-functors** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/221f33b5b7af4ebe1b331da167391c10b15622f9)
- [x] **modernize-use-uncaught-exceptions** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/221f33b5b7af4ebe1b331da167391c10b15622f9)
- [x] **modernize-use-using** (no issues found as of https://github.com/azerothcore/azerothcore-wotlk/commit/221f33b5b7af4ebe1b331da167391c10b15622f9)

**Comment by FrancescoBorzi** [comment]
@azerothcore/developers @azerothcore/testers  I'm finally done with this. I need reviewers/testers for all the linked PRs, which you can find here:

https://github.com/azerothcore/azerothcore-wotlk/pulls?q=is%3Apr+is%3Aopen+refactor%28Core%29%3A+apply+clang-tidy+modernize

**Comment by Kitzunu** [comment]
@FrancescoBorzi Status on modernize-use-auto

**Comment by FrancescoBorzi** [comment]
> @FrancescoBorzi Status on modernize-use-auto

that's the only one missing here

**Comment by FrancescoBorzi** [comment]
converted to Discussion as the initial work has been completed but this is still a good reference

**Comment by FrancescoBorzi** [comment]
# How-to

Requirement: install `clang-tidy` (for macOS see [https://stackoverflow.com/questions/53111082/how-to-install-clang-tidy-on-macos](this))

1) Run cmake with: `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON` it will generate a `compile_commands.json` file inside the build folder

2) Download [this](https://github.com/llvm-mirror/clang-tools-extra/blob/master/clang-tidy/tool/run-clang-tidy.py) and place it inside the build folder (where `compile_commands.json` is located)

3) Run the following:

```
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-concat-nested-namespaces' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-deprecated-headers' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-deprecated-ios-base-aliases' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-loop-convert' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-make-shared' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-make-unique' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-pass-by-value' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-raw-string-literal' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-redundant-void-arg' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-replace-auto-ptr' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-replace-disallow-copy-and-assign-macro' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-replace-random-shuffle' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-return-braced-init-list' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-shrink-to-fit' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-unary-static-assert' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-bool-literals' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-default-member-init' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-emplace' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-equals-default' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-equals-delete' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-nodiscard' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-noexcept' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-nullptr' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-override' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-trailing-return-type' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-transparent-functors' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-uncaught-exceptions' -fix;
 ./run-clang-tidy.py -header-filter='.*' -checks='-*,modernize-use-using' -fix;
```
